### PR TITLE
Add pre-commit hooks and Github action for linting and formatting

### LINKED ISSUE #167: Shoud the data/.tmp file be in github?

### PR DESCRIPTION
**Pull request by Wiebke**
This PR adds back the Github actions for linting and formatting that used to be part of [.github/workflows/staging-deploy.yml](https://github.com/mlexchange/mlex_highres_segmentation/blob/979d6e554289417a26ecfcb1f857ba4af9a7e6b6/.github/workflows/staging-deploy.yml) but were removed in #158 and adds and applies corresponding pre-commit hooks and settings for `black`, `isort`, and `flake8`, following [black's compatibility guidelines](https://black.readthedocs.io/en/stable/guides/using_black_with_other_tools.html).

### Additional Notes
The `test-and-lint` action is applied only on pull requests, not commits, and running the hooks thus can be omitted with the [`--no-verify` flag](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---no-verify) for commits with work in progress.

To setup pre-commit hooks run `pip install -r requirements-dev.txt`, followed by `pre-commit install`. Aside from during commits, the hooks can also be run with `pre-commit run --all-files`.

Some selected lines have been marked with `# noqa:` to [ignore flake8 errors](https://flake8.pycqa.org/en/3.0.1/user/ignoring-errors.html). These are the dash component imports as well as the long lines with cmd specifications. 

Additionally closes #167, as the sample data directory will be created if it doesn't exists when running `python3 utils/download_sample_data.py`

### CODE REVIEW

**Comment by Wiebke** [comment]
### Questions

`black` has a default line length of 88 characters. It will not auto-format comments, docstrings, or strings. The previous Github action in this repo had `flake8` check for a line length of 127 characters with the comment that is in line with the Github code editor. This means effectively since `black` runs before `flake8` in the pre-commit hooks,  `flake8` will only complain about long lines with comments, docstrings, or strings. Is that sensible?

I looked at Tiled's pre-commit hooks ([bluesky/tiled/blob/main/.pre-commit-config.yaml](https://github.com/bluesky/tiled/blob/main/.pre-commit-config.yaml)) as an example. There are a number not necessarily Python-related hooks that may be sensible to include as well, e.g. removing trailing white-space, uniform ending of files, warning about debug statements, etc. Should we also include these?

**Comment by dylanmcreynolds** [comment]
> There are a number not necessarily Python-related hooks that may be sensible to include as well, e.g. removing trailing white-space, uniform ending of files, warning about debug statements, etc. Should we also include these?

Yes. I think these are python, actually. `flake8` certainly complains about trailing white-space and uniform ending of files. I don't klnow if this will warn of `breakpoint()` statement, but it sure would be nice.

File: .github/workflows/test-lint.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  test-and-lint:
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Set up Python 3.9.16
+        uses: actions/setup-python@v3
+        with:
+          python-version: '3.9.16'
```
Comment (dylanmcreynolds) [inline .github/workflows/test-lint.yml:15]: Should we use 3.11?

File: requirements-dev.txt [hunk 2]
```diff
@@ -1,7 +1,8 @@
-black==22.10.0
-flake8==6.0.0
-tiled[all]
+black==24.2.0
+flake8==7.0.0
+tiled[all]==0.1.0a114
```
Comment (dylanmcreynolds) [inline requirements-dev.txt:3]: Do we need `tiled[all]` in this project? Would `tiled[client]` be sufficient?

**Review by dylanmcreynolds — COMMENTED** [review]
Looks good. I added a couple of minor comments.

**Review by dylanmcreynolds — COMMENTED** [review]
Looks good. I added a couple of minor comments.

File: requirements-dev.txt (hunk 2 above)
Comment (Wiebke) [reply requirements-dev.txt:3]: That's a good point, with a more complete test-setup in [mlexchange/mlex_tomo_framework](https://github.com/mlexchange/mlex_tomo_framework), it makes sense to remove this here.

**Review by dylanmcreynolds — APPROVED** [review]

**Review by taxe10 — APPROVED** [review]
All looks good to me!

File: .github/workflows/test-lint.yml (hunk 1 above)
Comment (Wiebke) [reply .github/workflows/test-lint.yml:15]: A higher version makes sense, yes.

However, noticed that our Docker image has `python:3.9` as a base image as well and thus would like to address that in a separate PR.

**Pull request merged**